qubes/devices.py

```diff
@@ -351,24 +353,19 @@ def assignments(self, persistent=None):
                 self._bus))
             if persistent is True:
                 # don't break app.save()
-                return self._set
+                return list(self._set)
             raise
-        result = set()
-        for dev, options in devices:
-            if dev in self._set and not persistent:
-                continue
-            if dev in self._set:
-                result.add(self._set.get(dev))
-            elif dev not in self._set and persistent:
-                continue
-            else:
-                result.add(
-                    DeviceAssignment(
-                        backend_domain=dev.backend_domain,
-                        ident=dev.ident, options=options,
-                        bus=self._bus))
-        if persistent is not False:
-            result.update(self._set)
+        result = []
+        if persistent is not False:  # None or True
+            result.extend(self._set)
+        if not persistent:  # None or False
+            for dev, options in devices:
+                if dev not in self._set:
+                    result.append(
+                        DeviceAssignment(
+                            backend_domain=dev.backend_domain,
+                            ident=dev.ident, options=options,
+                            bus=self._bus))
         return result
```

Comment:
Should this be `sorted(result)`?

Reply:
No. Otherwise adding new device could move another one attached earlier.